File: go/diacritics.go
package merkmal

import (
	"encoding/json"
	"fmt"
	"io/fs"
	"sort"
	"strconv"
	"strings"
)

// toneLevels holds Chao onset/mid/offset levels for a tone diacritic.
type toneLevels struct{ onset, mid, offset int }

// modifierEffect is the valued-feature change a modifier applies.
type modifierEffect struct {
	alternatives []string
	state        FeatureState
}

// DiacriticTable bundles the diacritic/modifier/tone → feature mappings
// and the decomposition logic that uses them. The feature *names* a
// modifier produces are part of a feature system's vocabulary, so the
// table is data-drivable: see LoadDiacritics. DefaultDiacritics is the
// built-in IPA/CLTS set.
type DiacriticTable struct {
	Name          string
	Combining     map[rune]string
	Suffix        map[rune]string
	Prefix        map[rune]string
	ToneMarks     map[rune]toneLevels
	ToneOnset     map[int][]string
	ToneMid       map[int][]string
	ToneOffset    map[int][]string
	ValuedEffects map[string]modifierEffect

	featureToModifier map[string]rune
}

func (d *DiacriticTable) reverseMap() map[string]rune {
	if d.featureToModifier != nil {
		return d.featureToModifier
	}
	result := make(map[string]rune)
	for cp, feat := range d.Combining {
		if _, exists := result[feat]; !exists {
			result[feat] = cp
		}
	}
	for cp, feat := range d.Suffix {
		result[feat] = cp
	}
	d.featureToModifier = result
	return result
}

// ToneFeaturesForLevels returns tone features for Chao onset/mid/offset levels.
func (d *DiacriticTable) ToneFeaturesForLevels(onset, mid, offset int) map[string]bool {
	result := map[string]bool{}
	for _, f := range d.ToneOnset[onset] {
		result[f] = true
	}
	for _, f := range d.ToneMid[mid] {
		result[f] = true
	}
	for _, f := range d.ToneOffset[offset] {
		result[f] = true
	}
	return result
}

// Decompose extracts base characters and modifier features from a grapheme.
func (d *DiacriticTable) Decompose(grapheme string) (base string, features map[string]bool) {
	runes := []rune(grapheme)
	features = map[string]bool{}

	prefixEnd := 0
	for i, r := range runes {
		if isLmOrSk(r) {
			if feat, ok := d.Prefix[r]; ok {
				features[feat] = true
				prefixEnd = i + 1
				continue
			}
		}
		break
	}

	remainder := runes[prefixEnd:]

	tailStart := len(remainder)
	for k := len(remainder) - 1; k >= 0; k-- {
		if isChaoDigit[remainder[k]] {
			tailStart = k
		} else {
			break
		}
	}

	var baseChars []rune
	var chaoChars []rune

	for idx, r := range remainder {
		if idx >= tailStart {
			chaoChars = append(chaoChars, r)
			continue
		}
		if levels, ok := d.ToneMarks[r]; ok {
			for f := range d.ToneFeaturesForLevels(levels.onset, levels.mid, levels.offset) {
				features[f] = true
			}
		} else if isMark(r) {
			if feat, ok := d.Combining[r]; ok {
				features[feat] = true
			} else {
				baseChars = append(baseChars, r)
			}
		} else if isLmOrSk(r) {
			if feat, ok := d.Suffix[r]; ok {
				features[feat] = true
			} else {
				baseChars = append(baseChars, r)
			}
		} else {
			baseChars = append(baseChars, r)
		}
	}

	if len(chaoChars) > 0 {
		o, m, off, ok := ParseChaoDigits(string(chaoChars))
		if ok {
			for f := range d.ToneFeaturesForLevels(o, m, off) {
				features[f] = true
			}
		}
	}

	return string(baseChars), features
}

// Compose reconstructs a grapheme from base + modifier features.
func (d *DiacriticTable) Compose(base string, modifiers map[string]bool) string {
	ftm := d.reverseMap()
	sorted := make([]string, 0, len(modifiers))
	for f := range modifiers {
		sorted = append(sorted, f)
	}
	sort.Strings(sorted)

	var b strings.Builder
	b.WriteString(base)
	for _, feat := range sorted {
		if ch, ok := ftm[feat]; ok {
			b.WriteRune(ch)
		}
	}
	return b.String()
}

// ApplyValuedEffects applies modifier effects to a copy of base valued features.
func (d *DiacriticTable) ApplyValuedEffects(baseValues map[string]FeatureState, modifiers, modelFeatures map[string]bool) map[string]FeatureState {
	result := make(map[string]FeatureState, len(baseValues))
	for k, v := range baseValues {
		result[k] = v
	}
	for modifier := range modifiers {
		effect, ok := d.ValuedEffects[modifier]
		if !ok {
			continue
		}
		for _, featName := range effect.alternatives {
			if modelFeatures[featName] {
				result[featName] = effect.state
				break
			}
		}
	}
	return result
}

// ── Built-in IPA / CLTS default ─────────────────────────────────────────

const defaultDiacriticsName = "ipa-clts"

func levelFeatures(prefix string) map[int][]string {
	return map[int][]string{
		5: {"tone-" + prefix + "-upper", "tone-" + prefix + "-raised"},
		4: {"tone-" + prefix + "-upper", "tone-" + prefix + "-lowered"},
		3: nil,
		2: {"tone-" + prefix + "-lower", "tone-" + prefix + "-raised"},
		1: {"tone-" + prefix + "-lower", "tone-" + prefix + "-lowered"},
	}
}

// DefaultDiacritics is the built-in IPA/CLTS diacritic table.
var DefaultDiacritics = &DiacriticTable{
	Name: defaultDiacriticsName,
	Combining: map[rune]string{
		0x0325: "devoiced",
		0x030A: "devoiced",
		0x032C: "revoiced",
		0x0330: "creaky",
		0x0324: "breathy",
		0x0303: "nasalized",
		0x0329: "syllabic",
		0x030D: "syllabic",
		0x032F: "non-syllabic",
		0x032A: "dental",
		0x031F: "advanced",
		0x0320: "retracted",
		0x0318: "advanced-tongue-root",
		0x0319: "retracted-tongue-root",
		0x033A: "apical",
		0x033B: "laminal",
		0x033C: "linguolabial",
		0x031D: "raised",
		0x031E: "lowered",
		0x0308: "centralized",
		0x033D: "mid-centralized",
		0x031C: "less-rounded",
		0x0339: "more-rounded",
		0x0306: "ultra-short",
		0x031A: "unreleased",
		0x0348: "strong",
	},
	Suffix: map[rune]string{
		0x02D0: "long",
		0x02D1: "mid-long",
		0x02B0: "aspirated",
		0x02B1: "breathy",
		0x02B2: "palatalized",
		0x02B7: "labialized",
		0x02E0: "velarized",
		0x02E4: "pharyngealized",
		0x02C0: "glottalized",
		0x02BC: "ejective",
		0x1DA3: "labio-palatalized",
		0x207F: "with-nasal-release",
		0x02DE: "rhotacized",
		0x02E1: "with-lateral-release",
	},
	Prefix: map[rune]string{
		0x02B0: "pre-aspirated",
		0x02C0: "pre-glottalized",
		0x207F: "pre-nasalized",
		0x02B7: "pre-labialized",
		0x02B2: "pre-palatalized",
	},
	ToneMarks: map[rune]toneLevels{
		0x030B: {5, 5, 5},
		0x0301: {4, 4, 4},
		0x0304: {3, 3, 3},
		0x0300: {2, 2, 2},
		0x030F: {1, 1, 1},
		0x0302: {4, 3, 2},
		0x030C: {2, 3, 4},
	},
	ToneOnset:  levelFeatures("onset"),
	ToneMid:    levelFeatures("mid"),
	ToneOffset: levelFeatures("offset"),
	ValuedEffects: map[string]modifierEffect{
		"devoiced":              {[]string{"periodicGlottalSource", "voice", "voiced"}, StateNegative},
		"revoiced":              {[]string{"periodicGlottalSource", "voice", "voiced"}, StatePositive},
		"aspirated":             {[]string{"spreadGlottis", "spread"}, StatePositive},
		"breathy":               {[]string{"spreadGlottis", "spread"}, StatePositive},
		"creaky":                {[]string{"constrictedGlottis", "constr"}, StatePositive},
		"nasalized":             {[]string{"nasal"}, StatePositive},
		"long":                  {[]string{"long", "LONG"}, StatePositive},
		"dental":                {[]string{"distributed"}, StatePositive},
		"syllabic":              {[]string{"syllabic", "SYLLABIC"}, StatePositive},
		"non-syllabic":          {[]string{"syllabic", "SYLLABIC"}, StateNegative},
		"ejective":              {[]string{"raisedLarynxEjective", "constrictedGlottis", "constr"}, StatePositive},
		"glottalized":           {[]string{"constrictedGlottis", "constr"}, StatePositive},
		"palatalized":           {[]string{"high"}, StatePositive},
		"labialized":            {[]string{"round"}, StatePositive},
		"more-rounded":          {[]string{"round"}, StatePositive},
		"less-rounded":          {[]string{"round"}, StateNegative},
		"velarized":             {[]string{"dorsal"}, StatePositive},
		"pharyngealized":        {[]string{"retractedTongueRoot"}, StatePositive},
		"advanced-tongue-root":  {[]string{"advancedTongueRoot", "ATR"}, StatePositive},
		"retracted-tongue-root": {[]string{"retractedTongueRoot"}, StatePositive},
	},
}

// ── Loading from JSON ────────────────────────────────────────────────────

type jsonDiacritics struct {
	Name          string                         `json:"name"`
	Combining     map[string]string              `json:"combining"`
	Suffix        map[string]string              `json:"suffix"`
	Prefix        map[string]string              `json:"prefix"`
	ToneMarks     map[string][]int               `json:"tone_marks"`
	ToneLevels    map[string]map[string][]string `json:"tone_levels"`
	ValuedEffects map[string]jsonValuedEffect    `json:"valued_effects"`
}

type jsonValuedEffect struct {
	Features []string `json:"features"`
	State    string   `json:"state"`
}

func parseCPMap(raw map[string]string) (map[rune]string, error) {
	out := make(map[rune]string, len(raw))
	for k, v := range raw {
		cp, err := strconv.ParseInt(k, 16, 32)
		if err != nil {
			return nil, fmt.Errorf("invalid codepoint %q: %w", k, err)
		}
		out[rune(cp)] = v
	}
	return out, nil
}

func parseLevels(raw map[string][]string) map[int][]string {
	out := make(map[int][]string, len(raw))
	for k, v := range raw {
		n, err := strconv.Atoi(k)
		if err != nil {
			continue
		}
		out[n] = v
	}
	return out
}

// LoadDiacritics reads a diacritic table JSON file from an fs.FS.
func LoadDiacritics(fsys fs.FS, name string) (*DiacriticTable, error) {
	data, err := fs.ReadFile(fsys, name+".json")
	if err != nil {
		return nil, fmt.Errorf("diacritics not found: %s (%w)", name, err)
	}
	var raw jsonDiacritics
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parsing diacritics %s: %w", name, err)
	}

	combining, err := parseCPMap(raw.Combining)
	if err != nil {
		return nil, err
	}
	suffix, err := parseCPMap(raw.Suffix)
	if err != nil {
		return nil, err
	}
	prefix, err := parseCPMap(raw.Prefix)
	if err != nil {
		return nil, err
	}

	toneMarks := make(map[rune]toneLevels, len(raw.ToneMarks))
	for k, v := range raw.ToneMarks {
		cp, err := strconv.ParseInt(k, 16, 32)
		if err != nil || len(v) != 3 {
			return nil, fmt.Errorf("invalid tone_mark %q", k)
		}
		toneMarks[rune(cp)] = toneLevels{v[0], v[1], v[2]}
	}

	effects := make(map[string]modifierEffect, len(raw.ValuedEffects))
	for mod, spec := range raw.ValuedEffects {
		state, err := parseStateSymbol(spec.State)
		if err != nil {
			return nil, fmt.Errorf("valued_effect %q: %w", mod, err)
		}
		effects[mod] = modifierEffect{alternatives: spec.Features, state: state}
	}

	d := &DiacriticTable{
		Name:          orDefault(raw.Name, name),
		Combining:     combining,
		Suffix:        suffix,
		Prefix:        prefix,
		ToneMarks:     toneMarks,
		ToneOnset:     parseLevels(raw.ToneLevels["onset"]),
		ToneMid:       parseLevels(raw.ToneLevels["mid"]),
		ToneOffset:    parseLevels(raw.ToneLevels["offset"]),
		ValuedEffects: effects,
	}
	return d, nil
}

func orDefault(s, fallback string) string {
	if s == "" {
		return fallback
	}
	return s
}

func parseStateSymbol(s string) (FeatureState, error) {
	switch s {
	case string(StatePositive):
		return StatePositive, nil
	case string(StateNegative):
		return StateNegative, nil
	default:
		// Allow word forms for readability.
		switch strings.ToLower(s) {
		case "positive", "+":
			return StatePositive, nil
		case "negative", "-":
			return StateNegative, nil
		}
		return "", fmt.Errorf("unknown feature state %q", s)
	}
}

File: tools/legacy_python/merkmal/diacritics.py
"""Diacritic / modifier / tone feature tables (data-drivable).

Diacritic composition turns a grapheme such as ``tʰ`` into a base
(``t``) plus modifier features (``aspirated``). Which feature *names*
those modifiers produce is part of a feature system's vocabulary, so it
must be configurable when a user brings their own model.

This module defines :class:`DiacriticTable`, which bundles the
codepoint→feature maps, the tone-level→feature maps, and the valued
modifier effects, together with the decomposition / composition logic
that uses them. :data:`DEFAULT_DIACRITICS` is the built-in IPA/CLTS set;
:func:`load_diacritics` loads alternative sets from
``diacritics/<name>.json`` on the layered data search path.

The Unicode *recognition* of modifiers (which codepoints are tie bars,
combining marks, etc.) and IPA input normalization remain global and
live in :mod:`merkmal.grapheme`; only the feature-name mapping is
parameterized here.
"""

from __future__ import annotations

import json
import unicodedata
from dataclasses import dataclass, field
from functools import cache
from typing import Any

from merkmal.representations import FeatureState
from merkmal.segmentation import parse_chao_digits

CHAO_SUPERSCRIPT_DIGITS: frozenset[str] = frozenset("⁰¹²³⁴⁵")


@dataclass
class DiacriticTable:
    """A named set of diacritic/modifier/tone → feature mappings."""

    name: str
    combining: dict[int, str]
    suffix: dict[int, str]
    prefix: dict[int, str]
    tone_marks: dict[int, tuple[int, int, int]]
    tone_onset: dict[int, frozenset[str]]
    tone_mid: dict[int, frozenset[str]]
    tone_offset: dict[int, frozenset[str]]
    valued_effects: dict[str, tuple[tuple[str, ...], FeatureState]]

    _feature_to_modifier: dict[str, str] = field(init=False, repr=False)

    def __post_init__(self) -> None:
        rev: dict[str, str] = {}
        for cp, feat in self.combining.items():
            rev.setdefault(feat, chr(cp))
        for cp, feat in self.suffix.items():
            rev[feat] = chr(cp)
        self._feature_to_modifier = rev

    # ── Tone ─────────────────────────────────────────────────────────
    def tone_features_for_levels(
        self, onset: int, mid: int, offset: int
    ) -> frozenset[str]:
        return (
            self.tone_onset[onset]
            | self.tone_mid[mid]
            | self.tone_offset[offset]
        )

    # ── Decomposition ────────────────────────────────────────────────
    def decompose(self, grapheme: str) -> tuple[str, frozenset[str]]:
        """Extract base characters and modifier features from a grapheme."""
        features: set[str] = set()

        prefix_end = 0
        for i, ch in enumerate(grapheme):
            cp = ord(ch)
            if unicodedata.category(ch) in ("Lm", "Sk") and cp in self.prefix:
                features.add(self.prefix[cp])
                prefix_end = i + 1
            else:
                break

        base_chars: list[str] = []
        chao_chars: list[str] = []
        remainder = grapheme[prefix_end:]

        tail_start = len(remainder)
        for k in range(len(remainder) - 1, -1, -1):
            if remainder[k] in CHAO_SUPERSCRIPT_DIGITS:
                tail_start = k
            else:
                break

        for idx, ch in enumerate(remainder):
            if idx >= tail_start:
                chao_chars.append(ch)
                continue
            cp = ord(ch)
            cat = unicodedata.category(ch)
            if cp in self.tone_marks:
                onset, mid, offset = self.tone_marks[cp]
                features |= self.tone_features_for_levels(onset, mid, offset)
            elif cat.startswith("M") and cp in self.combining:
                features.add(self.combining[cp])
            elif cat in ("Lm", "Sk") and cp in self.suffix:
                features.add(self.suffix[cp])
            else:
                base_chars.append(ch)

        if chao_chars:
            parsed = parse_chao_digits("".join(chao_chars))
            if parsed is not None:
                onset, mid, offset = parsed
                features |= self.tone_features_for_levels(onset, mid, offset)

        return "".join(base_chars), frozenset(features)

    # ── Composition ──────────────────────────────────────────────────
    def available_modifiers(self) -> tuple[str, ...]:
        return tuple(sorted(self._feature_to_modifier))

    def compose(self, base: str, modifiers: frozenset[str]) -> str:
        """Reconstruct a grapheme from base + modifier features."""
        suffix: list[str] = []
        for feat in sorted(modifiers):
            char = self._feature_to_modifier.get(feat)
            if char is not None:
                suffix.append(char)
        return base + "".join(suffix)

    # ── Valued effects ───────────────────────────────────────────────
    def apply_valued_effects(
        self,
        base_values: dict[str, FeatureState],
        modifiers: frozenset[str],
        model_features: set[str],
    ) -> dict[str, FeatureState]:
        """Apply modifier effects to a copy of base valued features.

        For each modifier, looks up the first matching feature name from
        the model's feature set and sets it to the target state.
        """
        result = dict(base_values)
        for modifier in modifiers:
            effect = self.valued_effects.get(modifier)
            if effect is None:
                continue
            alternatives, target_state = effect
            for feat_name in alternatives:
                if feat_name in model_features:
                    result[feat_name] = target_state
                    break
        return result


# ── Built-in IPA / CLTS default ─────────────────────────────────────────

_DEFAULT_COMBINING: dict[int, str] = {
    0x0325: "devoiced",
    0x030A: "devoiced",
    0x032C: "revoiced",
    0x0330: "creaky",
    0x0324: "breathy",
    0x0303: "nasalized",
    0x0329: "syllabic",
    0x030D: "syllabic",
    0x032F: "non-syllabic",
    0x032A: "dental",
    0x031F: "advanced",
    0x0320: "retracted",
    0x0318: "advanced-tongue-root",
    0x0319: "retracted-tongue-root",
    0x033A: "apical",
    0x033B: "laminal",
    0x033C: "linguolabial",
    0x031D: "raised",
    0x031E: "lowered",
    0x0308: "centralized",
    0x033D: "mid-centralized",
    0x031C: "less-rounded",
    0x0339: "more-rounded",
    0x0306: "ultra-short",
    0x031A: "unreleased",
    0x0348: "strong",
}

_DEFAULT_SUFFIX: dict[int, str] = {
    0x02D0: "long",
    0x02D1: "mid-long",
    0x02B0: "aspirated",
    0x02B1: "breathy",
    0x02B2: "palatalized",
    0x02B7: "labialized",
    0x02E0: "velarized",
    0x02E4: "pharyngealized",
    0x02C0: "glottalized",
    0x02BC: "ejective",
    0x1DA3: "labio-palatalized",
    0x207F: "with-nasal-release",
    0x02DE: "rhotacized",
    0x02E1: "with-lateral-release",
}

_DEFAULT_PREFIX: dict[int, str] = {
    0x02B0: "pre-aspirated",
    0x02C0: "pre-glottalized",
    0x207F: "pre-nasalized",
    0x02B7: "pre-labialized",
    0x02B2: "pre-palatalized",
}

_DEFAULT_TONE_MARKS: dict[int, tuple[int, int, int]] = {
    0x030B: (5, 5, 5),
    0x0301: (4, 4, 4),
    0x0304: (3, 3, 3),
    0x0300: (2, 2, 2),
    0x030F: (1, 1, 1),
    0x0302: (4, 3, 2),
    0x030C: (2, 3, 4),
}


def _tone_levels(prefix: str) -> dict[int, frozenset[str]]:
    return {
        5: frozenset({f"tone-{prefix}-upper", f"tone-{prefix}-raised"}),
        4: frozenset({f"tone-{prefix}-upper", f"tone-{prefix}-lowered"}),
        3: frozenset(),
        2: frozenset({f"tone-{prefix}-lower", f"tone-{prefix}-raised"}),
        1: frozenset({f"tone-{prefix}-lower", f"tone-{prefix}-lowered"}),
    }


_DEFAULT_VALUED_EFFECTS: dict[str, tuple[tuple[str, ...], FeatureState]] = {
    "devoiced": (("periodicGlottalSource", "voice", "voiced"), FeatureState.NEGATIVE),
    "revoiced": (("periodicGlottalSource", "voice", "voiced"), FeatureState.POSITIVE),
    "aspirated": (("spreadGlottis", "spread"), FeatureState.POSITIVE),
    "breathy": (("spreadGlottis", "spread"), FeatureState.POSITIVE),
    "creaky": (("constrictedGlottis", "constr"), FeatureState.POSITIVE),
    "nasalized": (("nasal",), FeatureState.POSITIVE),
    "long": (("long", "LONG"), FeatureState.POSITIVE),
    "dental": (("distributed",), FeatureState.POSITIVE),
    "syllabic": (("syllabic", "SYLLABIC"), FeatureState.POSITIVE),
    "non-syllabic": (("syllabic", "SYLLABIC"), FeatureState.NEGATIVE),
    "ejective": (
        ("raisedLarynxEjective", "constrictedGlottis", "constr"),
        FeatureState.POSITIVE,
    ),
    "glottalized": (("constrictedGlottis", "constr"), FeatureState.POSITIVE),
    "palatalized": (("high",), FeatureState.POSITIVE),
    "labialized": (("round",), FeatureState.POSITIVE),
    "more-rounded": (("round",), FeatureState.POSITIVE),
    "less-rounded": (("round",), FeatureState.NEGATIVE),
    "velarized": (("dorsal",), FeatureState.POSITIVE),
    "pharyngealized": (("retractedTongueRoot",), FeatureState.POSITIVE),
    "advanced-tongue-root": (("advancedTongueRoot", "ATR"), FeatureState.POSITIVE),
    "retracted-tongue-root": (("retractedTongueRoot",), FeatureState.POSITIVE),
}

DEFAULT_DIACRITICS_NAME = "ipa-clts"

DEFAULT_DIACRITICS = DiacriticTable(
    name=DEFAULT_DIACRITICS_NAME,
    combining=dict(_DEFAULT_COMBINING),
    suffix=dict(_DEFAULT_SUFFIX),
    prefix=dict(_DEFAULT_PREFIX),
    tone_marks=dict(_DEFAULT_TONE_MARKS),
    tone_onset=_tone_levels("onset"),
    tone_mid=_tone_levels("mid"),
    tone_offset=_tone_levels("offset"),
    valued_effects=dict(_DEFAULT_VALUED_EFFECTS),
)


# ── Loading from JSON ────────────────────────────────────────────────────

def _parse_cp_map(raw: dict[str, str]) -> dict[int, str]:
    return {int(k, 16): v for k, v in raw.items()}


def _parse_levels(raw: dict[str, list[str]]) -> dict[int, frozenset[str]]:
    return {int(k): frozenset(v) for k, v in raw.items()}


def _parse_valued_effects(
    raw: dict[str, Any],
) -> dict[str, tuple[tuple[str, ...], FeatureState]]:
    out: dict[str, tuple[tuple[str, ...], FeatureState]] = {}
    for modifier, spec in raw.items():
        feats = tuple(spec["features"])
        out[modifier] = (feats, FeatureState(spec["state"]))
    return out


def parse_diacritics(data: dict[str, Any], name: str) -> DiacriticTable:
    """Build a :class:`DiacriticTable` from parsed JSON data."""
    tone = data.get("tone_levels", {})
    return DiacriticTable(
        name=data.get("name", name),
        combining=_parse_cp_map(data.get("combining", {})),
        suffix=_parse_cp_map(data.get("suffix", {})),
        prefix=_parse_cp_map(data.get("prefix", {})),
        tone_marks={
            int(k, 16): tuple(v) for k, v in data.get("tone_marks", {}).items()
        },
        tone_onset=_parse_levels(tone.get("onset", {})),
        tone_mid=_parse_levels(tone.get("mid", {})),
        tone_offset=_parse_levels(tone.get("offset", {})),
        valued_effects=_parse_valued_effects(data.get("valued_effects", {})),
    )


@cache
def load_diacritics(name: str | None) -> DiacriticTable:
    """Load a diacritic table by name from the layered search path.

    ``None`` or the default name returns the built-in IPA/CLTS table
    unless an override file of that name is found on the search path.
    """
    from merkmal import paths

    if not name:
        return DEFAULT_DIACRITICS
    path = paths.resolve_file("diacritics", f"{name}.json")
    if path is None:
        if name == DEFAULT_DIACRITICS_NAME:
            return DEFAULT_DIACRITICS
        roots = paths.data_roots("diacritics")
        msg = f"Diacritic set not found: {name} (looked in {roots})"
        raise FileNotFoundError(msg)
    data = json.loads(path.read_text(encoding="utf-8"))
    return parse_diacritics(data, name)
